File: server/llm-analyzer.ts
import OpenAI from "openai";
import type { Claim, InsertLlmSignal } from "@shared/schema";

const openai = new OpenAI({
  apiKey: process.env.AI_INTEGRATIONS_OPENAI_API_KEY,
  baseURL: process.env.AI_INTEGRATIONS_OPENAI_BASE_URL,
});

const ANALYSIS_PROMPT = `You are a neutral fraud analysis assistant for UK motor insurance claims. Your role is to identify potential inconsistencies, anomalies, or noteworthy patterns in claim data. You do NOT determine if something is fraudulent - you only highlight observations that may warrant further human investigation.

CRITICAL RULES:
1. Use neutral, objective language only. Never use words like "suspicious", "fraudulent", "deceptive", "dishonest", or similar judgmental terms.
2. Focus on factual observations: date mismatches, unusual values, inconsistencies in narrative.
3. For each observation, cite specific evidence from the data.
4. Provide a confidence level (0.0 to 1.0) for each observation.
5. If the claim appears straightforward, return fewer signals.

TYPES OF SIGNALS TO LOOK FOR:
- Date inconsistencies (accident date vs claim submission timing)
- Repair cost anomalies (amounts significantly above/below vehicle value)
- Description inconsistencies (narrative contradictions)
- History patterns (frequent claims, high cumulative amounts)
- Location/timing patterns (unusual accident circumstances)
- Documentation gaps (missing expected details)

OUTPUT FORMAT:
Return a JSON array of signals. Each signal should have:
- signalType: A short category name (e.g., "Date Mismatch", "Cost Anomaly", "Description Gap")
- description: A neutral description of the observation
- evidence: The specific data that led to this observation
- confidence: A number from 0.0 to 1.0

Example output:
[
  {
    "signalType": "Repair Cost Note",
    "description": "The claimed repair amount represents a significant proportion of the vehicle's estimated value",
    "evidence": "Claim amount: £8,500 for vehicle valued at £12,000 (71% of value)",
    "confidence": 0.75
  }
]

If no noteworthy patterns are found, return an empty array: []`;

export interface AnalysisSignal {
  signalType: string;
  description: string;
  evidence: string;
  confidence: number;
}

export async function analyzeClaim(claim: Claim): Promise<AnalysisSignal[]> {
  const vehicleDetails = claim.vehicleDetails as { make: string; model: string; year: number; registration: string; estimatedValue: number };
  const claimantHistory = claim.claimantHistory as { previousClaims: number; lastClaimDate: string | null; totalPreviousAmount: number };

  const claimData = `
CLAIM DATA FOR ANALYSIS:
========================
Claim Reference: ${claim.claimRef}
Claim Amount: £${Number(claim.claimAmount).toLocaleString()}
Accident Date: ${new Date(claim.accidentDate).toDateString()}
Claim Submitted: ${new Date(claim.createdAt).toDateString()}
Accident Location: ${claim.accidentLocation}
Accident Type: ${claim.accidentType}

VEHICLE INFORMATION:
- Make/Model: ${vehicleDetails.year} ${vehicleDetails.make} ${vehicleDetails.model}
- Registration: ${vehicleDetails.registration}
- Estimated Value: £${vehicleDetails.estimatedValue.toLocaleString()}

CLAIMANT HISTORY:
- Previous Claims: ${claimantHistory.previousClaims}
- Total Previous Amount: £${claimantHistory.totalPreviousAmount.toLocaleString()}
- Last Claim Date: ${claimantHistory.lastClaimDate || "None"}

ACCIDENT DESCRIPTION:
${claim.accidentDescription}
`;

  try {
    const response = await openai.chat.completions.create({
      model: "gpt-5-mini",
      messages: [
        { role: "system", content: ANALYSIS_PROMPT },
        { role: "user", content: claimData },
      ],
      response_format: { type: "json_object" },
      max_completion_tokens: 1024,
    });

    const content = response.choices[0]?.message?.content;
    if (!content) return [];

    const parsed = JSON.parse(content);
    const signals = Array.isArray(parsed) ? parsed : (parsed.signals || []);
    
    return signals.filter((s: any) => 
      s.signalType && 
      s.description && 
      s.evidence && 
      typeof s.confidence === 'number'
    );
  } catch (error) {
    console.error("LLM analysis error:", error);
    return [];
  }
}

export function convertToInsertSignals(claimId: number, signals: AnalysisSignal[]): InsertLlmSignal[] {
  return signals.map(signal => ({
    claimId,
    signalType: signal.signalType,
    description: signal.description,
    evidence: signal.evidence,
    confidence: signal.confidence.toFixed(2),
    sourceDocument: null,
  }));
}

File: backend/app/services/llm_analyzer.py
import os
import json
from typing import List, Dict, Any, Optional
from datetime import datetime
import uuid
from openai import AzureOpenAI

def get_azure_openai_client() -> AzureOpenAI:
    endpoint = os.environ.get("AZURE_OPENAI_ENDPOINT")
    api_key = os.environ.get("AZURE_OPENAI_KEY")
    
    if not endpoint or not api_key:
        raise ValueError("Azure OpenAI credentials not configured")
    
    return AzureOpenAI(
        azure_endpoint=endpoint,
        api_key=api_key,
        api_version="2024-02-15-preview"
    )

SIGNAL_DETECTION_PROMPT = '''You are a neutral claims analysis assistant for UK motor insurance.
Analyze the following claim details and identify any factual observations that may be relevant for review.

IMPORTANT GUIDELINES:
- Use ONLY neutral, non-judgmental language
- Do NOT use words like: suspicious, fraudulent, deceptive, dishonest, false, fake, or scam
- Focus on factual observations and data patterns only
- Each observation should describe what you notice, not what you suspect

Claim Details:
- Claimant: {claimant_name}
- Policy ID: {policy_id}
- Previous Claims: {num_previous_claims} (Total: £{total_previous_claims_gbp})
- Vehicle: {vehicle_year} {vehicle_make} {vehicle_model} ({vehicle_registration})
- Vehicle Value: £{vehicle_estimated_value_gbp}
- Accident Date: {accident_date}
- Accident Type: {accident_type}
- Location: {accident_location}
- Claim Amount: £{claim_amount_gbp}
- Description: {accident_description}

Provide 0-5 observations in this exact JSON format:
{{
  "signals": [
    {{
      "signal_type": "Category (e.g., Cost Analysis, Timeline Observation, Documentation Gap, Pattern Note)",
      "description": "Neutral factual observation",
      "confidence": 0.0-1.0
    }}
  ],
  "summary": "Brief neutral summary of observations"
}}

Return ONLY valid JSON, no other text.'''

async def analyze_claim_signals(claim: Dict[str, Any]) -> List[Dict[str, Any]]:
    try:
        client = get_azure_openai_client()
        
        prompt = SIGNAL_DETECTION_PROMPT.format(
            claimant_name=claim.get("claimant_name", "Unknown"),
            policy_id=claim.get("policy_id", "Unknown"),
            num_previous_claims=claim.get("num_previous_claims", 0),
            total_previous_claims_gbp=claim.get("total_previous_claims_gbp", 0),
            vehicle_year=claim.get("vehicle_year", "Unknown"),
            vehicle_make=claim.get("vehicle_make", "Unknown"),
            vehicle_model=claim.get("vehicle_model", "Unknown"),
            vehicle_registration=claim.get("vehicle_registration", "Unknown"),
            vehicle_estimated_value_gbp=claim.get("vehicle_estimated_value_gbp", 0),
            accident_date=claim.get("accident_date", "Unknown"),
            accident_type=claim.get("accident_type", "Unknown"),
            accident_location=claim.get("accident_location", "Unknown"),
            claim_amount_gbp=claim.get("claim_amount_gbp", 0),
            accident_description=claim.get("accident_description", "No description provided")
        )
        
        response = client.chat.completions.create(
            model="gpt-4.1",
            messages=[
                {"role": "system", "content": "You are a neutral claims analysis assistant. Respond only with valid JSON."},
                {"role": "user", "content": prompt}
            ],
            temperature=0.3,
            max_tokens=1000
        )
        
        content = response.choices[0].message.content.strip()
        if content.startswith("```"):
            content = content.split("```")[1]
            if content.startswith("json"):
                content = content[4:]
        
        result = json.loads(content)
        signals = result.get("signals", [])
        
        formatted_signals = []
        for signal in signals:
            formatted_signals.append({
                "id": str(uuid.uuid4()),
                "signal_type": signal.get("signal_type", "General Observation"),
                "description": signal.get("description", ""),
                "confidence": float(signal.get("confidence", 0.5)),
                "detected_at": datetime.utcnow().isoformat()
            })
        
        return formatted_signals
        
    except Exception as e:
        print(f"Error analyzing claim: {e}")
        return []
